gaphor/UML/actions/partition.py

```diff
@@ -1,154 +1,139 @@
 
 from gaphor import UML
-from gaphor.core.styling import VerticalAlign
-from gaphor.diagram.presentation import ElementPresentation, Named
-from gaphor.diagram.shapes import Box, Text, cairo_state, draw_highlight, stroke
+from gaphor.core.modeling import DrawContext
+from gaphor.core.modeling.properties import association, attribute
+from gaphor.core.styling import Style, VerticalAlign
+from gaphor.diagram.presentation import ElementPresentation
+from gaphor.diagram.shapes import Box, cairo_state, draw_highlight, stroke
 from gaphor.diagram.support import represents
-from gaphor.UML.modelfactory import stereotypes_str
-
+from gaphor.diagram.text import Layout
 
-@represents(UML.ActivityPartition)
-class PartitionItem(ElementPresentation, Named):
+INIT_NUM_PARTITIONS: int = 2
+HEADER_HEIGHT: int = 29
 
-    DELTA = 30
 
+@represents(UML.ActivityPartition)
+class PartitionItem(ElementPresentation):
     def __init__(self, id=None, model=None):
         super().__init__(id, model)
-        self._toplevel = False
-        self._bottom = False
-        self._subpart = False
-        self._hdmax = 0  # maximum subpartition header height
-
-        self.shape = Box(
-            Text(
-                text=lambda: stereotypes_str(
-                    self.subject,
-                    self.subject and self.subject.isExternal and ("external",) or (),
-                ),
-            ),
-            Text(text=lambda: self.subject.name or ""),
-            style={
-                "min-width": 0,
-                "min-height": 0,
-                "line-width": 2.4,
-                "vertical-align": VerticalAlign.TOP,
-                "padding": (2, 2, 2, 2),
-            },
-            draw=self.draw_partition,
-        )
-        self.min_width = 100
         self.min_height = 300
+        self.style: Style = {
+            "font-family": "sans",
+            "font-size": 14,
+            "line-width": 2.4,
+            "padding": (2, 2, 2, 2),
+            "vertical-align": VerticalAlign.TOP,
+        }
 
         self.watch("subject[NamedElement].name")
         self.watch("subject.appliedStereotype.classifier.name")
+        self.watch("partition.name", self.update_shapes)
+        self.watch("num_partitions", self.update_shapes)
 
-    @property
-    def toplevel(self):
-        return self._toplevel
-
-    def pre_update(self, context):
-        assert self.canvas
-
-        self._header_size = self.shape.size(context)
-
-        # get subpartitions
-        children: List[PartitionItem] = [
-            k for k in self.canvas.get_children(self) if isinstance(k, PartitionItem)
-        ]
-
-        self._toplevel = self.canvas.get_parent(self) is None
-        self._subpart = len(children) > 0
-        self._bottom = not (self._toplevel or self._subpart)
-
-        if self._toplevel:
-            self._header_size = self._header_size[0], self.DELTA
-
-        handles = self.handles()
+    num_partitions: attribute[int] = attribute(
```

Comment:
With the partition association at hand, we should be able to get rid of this counter.